### PR TITLE
Add support for providing passkey to SMP pairing process

### PR DESCRIPTION
**Pull request by jkstrick-k4**

### CODE REVIEW

File: lib/hci-socket/smp.js [hunk 1]
```diff
@@ -28,14 +29,21 @@ var Smp = function(aclStream, localAddressType, localAddress, remoteAddressType,
 
   this._aclStream.on('data', this.onAclStreamDataBinded);
   this._aclStream.on('end', this.onAclStreamEndBinded);
+
+  this._remoteAddress = remoteAddress;
```
Comment (CTunstall) [inline lib/hci-socket/smp.js:33]: Is this._remoteAddress necessary, when we already have this._ra, which is at least somewhat similar?

File: lib/hci-socket/smp.js [hunk 2]
```diff
@@ -28,14 +29,21 @@ var Smp = function(aclStream, localAddressType, localAddress, remoteAddressType,
...
+    this._passkeyProvider = Promise.resolve;
 };
 
 util.inherits(Smp, events.EventEmitter);
 
 Smp.prototype.sendPairingRequest = function() {
   this._preq = new Buffer([
     SMP_PAIRING_REQUEST,
-    0x03, // IO capability: NoInputNoOutput
+    0x04, // IO capability: KeyboardDisplay
```
Comment (CTunstall) [inline lib/hci-socket/smp.js:46]: Did this have to change in order for a passkey to be allowed?

File: lib/hci-socket/gatt.js [hunk 3]
```diff
@@ -153,7 +153,9 @@ Gatt.prototype.onAclStreamEncrypt = function(encrypt) {
   if (encrypt) {
     this._security = 'medium';
 
-    this.writeAtt(this._currentCommand.buffer);
+    if(this._currentCommand) {
```
Comment (CTunstall) [inline lib/hci-socket/gatt.js:156]: I assume during testing that you ran into cases where where onAclStreamEncrypt was called, but this._currentCommand was null, which through an error when trying to write its buffer?

File: lib/hci-socket/smp.js (hunk 1 above)
Comment (jkstrick-k4) [reply lib/hci-socket/smp.js:33]: this._ra is the raw binary data that gets passed into the crypto function, and is not really useful for passing along to passkeyProvider().  We could convert it back from the buffer when we need it, but that seems like more trouble than it's worth when we have (almost) what we need already in the remoteAddress parameter.

File: lib/hci-socket/smp.js (hunk 2 above)
Comment (jkstrick-k4) [reply lib/hci-socket/smp.js:46]: Yes

File: lib/hci-socket/gatt.js (hunk 3 above)
Comment (jkstrick-k4) [reply lib/hci-socket/gatt.js:156]: That is correct.

**Review by CTunstall — DISMISSED** [review]

File: lib/hci-socket/bindings.js [hunk 4]
```diff
@@ -508,4 +510,8 @@ NobleBindings.prototype.onConnectionParameterUpdateRequest = function(handle, mi
   this._hci.connUpdateLe(handle, minInterval, maxInterval, latency, supervisionTimeout);
 };
 
+NobleBindings.prototype.registerPasskeyProvider = function(passkeyProvider) {
```
Comment (fantareinos) [inline lib/hci-socket/bindings.js:513]: ~~Isn't there also a definition for `registerPasskeyProvider` in `lib/noble.js`? 
What is the significance of having one here?~~

Never mind

File: lib/hci-socket/smp.js [hunk 5]
```diff
@@ -28,14 +29,21 @@ var Smp = function(aclStream, localAddressType, localAddress, remoteAddressType,
 
   this._aclStream.on('data', this.onAclStreamDataBinded);
   this._aclStream.on('end', this.onAclStreamEndBinded);
+
+  this._remoteAddress = remoteAddress;
+
+  if(passkeyProvider)
+    this._passkeyProvider = passkeyProvider;
+  else
+    this._passkeyProvider = Promise.resolve;
```
Comment (fantareinos) [inline lib/hci-socket/smp.js]: Line 38 should be `this._passkeyProvider = Promise.resolve();` right? Unless you want a function pointer...

Comment (fantareinos) [reply lib/hci-socket/smp.js]: I tried mimicking `this._passkeyProvider(this._remoteAddress.toLowerCase().replace(/\:/g, ''))` etc. by executing: 
```
var mockProvider = Promise.resolve; 
var mockAddr = "addressString"; 
mockProvider(mockAddr)
    .then((mockPassKey) => {
        console.log("Success");
    })
    .catch(err => console.log(err));
```

This gave me a: `TypeError: PromiseResolve called on non-object
    at resolve (<anonymous>)`

Comment (jkstrick-k4) [reply lib/hci-socket/smp.js]: Good catch.  Also very odd.  What kind of sense does this make anyway?
```
> var mockProvider = Promise.resolve
undefined
> mockProvider
[Function: resolve]
> mockProvider()
TypeError: mockProvider is not a function
    at resolve (native)
    at repl:1:1
    at REPLServer.defaultEval (repl.js:248:27)
    at bound (domain.js:280:14)
    at REPLServer.runBound [as eval] (domain.js:293:12)
    at REPLServer.<anonymous> (repl.js:412:12)
    at emitOne (events.js:82:20)
    at REPLServer.emit (events.js:169:7)
    at REPLServer.Interface._onLine (readline.js:210:10)
    at REPLServer.Interface._line (readline.js:549:8)
> mockProvider = function resolve() {}
[Function: resolve]
> mockProvider
[Function: resolve]
> mockProvider()
undefined
```

Comment (jkstrick-k4) [reply lib/hci-socket/smp.js]: I did actually intend to set it to the function reference, but that appears not to work.

**Review by CTunstall — APPROVED** [review]

**Review by fantareinos — APPROVED** [review]

**Review by CTunstall — APPROVED** [review]

**Pull request merged**